src/aspire/ctf/ctf_estimator.py

```diff
@@ -0,0 +1,757 @@
+            monotone_condition[np.arange(N - 1), np.arange(N - 1)] = -1
+            monotone_condition[np.arange(N - 1), np.arange(1, N)] = 1
+            monotone_condition = np.concatenate(
+                (np.zeros((N, N)), monotone_condition), axis=1
+            )
+
+            convex_condition = np.zeros((N, N))
+            convex_condition[np.arange(N - 2), np.arange(N - 2)] = -1
+            convex_condition[np.arange(N - 2), np.arange(1, N - 1)] = 2
+            convex_condition[np.arange(N - 2), np.arange(2, N)] = -1
+            convex_condition = np.concatenate(
+                (np.zeros((N, N)), convex_condition), axis=1
+            )
+
+            positivity_condition = np.concatenate(
+                (np.zeros((N, N)), -1 * np.eye(N)), axis=1
+            )
+
+            A = np.concatenate(
+                (superposition_condition, convex_condition, positivity_condition),
+                axis=0,
+            )
+
+            x_bound_lst = [
+                (signal[i], signal[i], -1 * np.inf, np.inf)
+                for i in range(signal.shape[0])
+            ]
+            x_bound = np.asarray(x_bound_lst, A.dtype)
+            x_bound = np.concatenate((x_bound[:, :2], x_bound[:, 2:]), axis=0)
+
+            x = linprog(
+                f,
+                A_ub=A,
+                b_ub=np.zeros((A.shape[0])),
+                bounds=x_bound,
+                method=linprog_method,
+            )
+            background = x.x[N:]
+
+            bs_psd = signal - background
+
+            final_signal[m:, m - 1] = bs_psd
+            final_background[m:, m - 1] = background  # difference: 10^-7 (absolute)
+
+        return final_signal, final_background
+
+    def ctf_opt1d(self, thon_rings, pixel_size, cs, lmbd, w, N):
+        """
+        Find optimal defocus for the radially symmetric case (where no astigmatism is present)
+
+        :param thon_rings: Estimated power specrtum.
+        :param pixel_size: Pixel size in Angstrom
+        :param cs: Spherical aberration in mm.
+        :param lmbd: Electron wavelength.
+        :param w: Amplitude contrast.
+        :param N: Number of rows (or columns) in the estimate power spectrum.
+        :return: 2-tuple of numpy arrays (Estimated average of defocus and an internal variable)
+        """
+
+        center = N // 2
+        [X, Y] = np.meshgrid(
+            np.arange(0 - center, N - center, dtype=self.dtype) / N,
+            np.arange(0 - center, N - center, dtype=self.dtype) / N,
+        )
+
+        rb = np.sqrt(np.square(X) + np.square(Y))
+        rb = rb[center, center:]
+        r_ctf = rb * (10 / pixel_size)
+
+        signal = thon_rings
+        signal = np.where(signal < 0, 0, signal)
+        signal = np.sqrt(signal)
+        signal = signal[: 3 * signal.shape[0] // 4]
+
+        r_ctf_sq = np.square(r_ctf)
+        c = np.zeros((9500, 13), dtype=self.dtype)
+
+        for f in range(500, 10000):
+            ctf_im = np.abs(
```

Comment:
Im not sure. I feel like because the main feature of the contributed code is supposed to be CTF, if we are/aren't implementing the CTF as ASPIRE needs it, that should go back to Ayelet.  I agree that we shoudn't have mutiple copy pasta blocks of the same sort of code, but I couldn't tell you which one is which... or if it is equivalent to what ASPIRE already has... and there isn't really much (any) documentation for me to sort it out even if I wanted to.

Reply:
#416